README.md
<h1 align="center" >Py Sberbank Payments</h1>

<p align="center">
	<a href="https://pypi.org/project/py_sberbank_payments">
		<img src="https://img.shields.io/pypi/status/py_sberbank_payments.svg" alt="PyPI - Status" />
	</a>
	<a href="https://pypi.org/project/py_sberbank_payments">
		<img src="https://img.shields.io/pypi/v/py_sberbank_payments.svg" alt="PyPI - Version" />
	</a>
	<a href="https://pypi.org/project/py_sberbank_payments">
		<img src="https://img.shields.io/pypi/dm/py_sberbank_payments.svg" alt="PyPI - Downloads" />
	</a>
	<a href="https://pypi.org/project/py_sberbank_payments">
		<img src="https://img.shields.io/pypi/pyversions/py_sberbank_payments.svg" alt="PyPI - Python Version" />
	</a>
	<a href="https://pypi.org/project/py_sberbank_payments">
		<img src="https://img.shields.io/pypi/format/py_sberbank_payments.svg" alt="PyPI - Format" />
	</a>
</p>
<p align="center">
	<a href="https://github.com/gocream/py_sberbank_payments/actions/workflows/mkdocs-release.yml" >
		<img src="https://img.shields.io/github/actions/workflow/status/gocream/py_sberbank_payments/mkdocs-release.yml?logo=github&label=docs" alt="Documentation - Release" />
	</a>
	<!-- <a href="https://github.com/gocream/py_sberbank_payments/actions/workflows/test.yml" >
		<img src="https://img.shields.io/github/actions/workflow/status/gocream/py_sberbank_payments/test.yml?logo=github&label=tests" alt="Tests - Running" />
	</a> -->
	<!-- <a href="https://codecov.io/gh/gocream/py_sberbank_payments" >
		<img src="https://codecov.io/gh/gocream/py_sberbank_payments/branch/main/graph/badge.svg?token=HV1QGD74EK" alt="Tests - Coverage" />
	</a> -->
</p>
<p align="center">
	<a href="https://github.com/pypa/hatch" target="_blank">
		<img src="https://img.shields.io/badge/%F0%9F%A5%9A-Hatch-4051b5.svg" alt="Hatch project" />
	</a>
	<a href="https://squidfunk.github.io/mkdocs-material/" target="_blank">
		<img src="https://img.shields.io/badge/docs-mkdocs_material-blue?logo=mdbook&logoColor=white" alt="Material for MkDocs" />
	</a>
	<a href="https://github.com/charliermarsh/ruff" target="_blank">
		<img src="https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/charliermarsh/ruff/main/assets/badge/v2.json" alt="linting - Ruff" />
	</a>
	<a href="https://github.com/psf/black" target="_blank">
		<img src="https://img.shields.io/badge/code%20style-black-000000.svg" alt="code style - black" />
	</a>
	<a href="https://raw.githubusercontent.com/gocream/py_sberbank_payments/master/LICENSE" target="_blank">
		<img src="https://img.shields.io/pypi/l/py_sberbank_payments?color=008033" alt="License - GNU Lesser General Public License v3.0" />
	</a>
</p>

Python wrapper for [Sberbank Payments API](https://securepayments.sberbank.ru/wiki/doku.php/start).

Тестовые карты - [link][test_cards]


## Development

`pip install -e .[dev]`


## TO DO:

* all errors inheritance SberbankException and raise them
* add all params from api to class methods arguments
* unittests
* test flask app


[test_cards]: https://securepayments.sberbank.ru/wiki/doku.php/test_cards


CHANGELOG.md
# Changelog

<a name="next"></a>
## next

### Changed

- 🔧 (lint): Linter configured [[6bb8e43](https://gitlab.com/gocream/pysberbank/commit/6bb8e43511f70c40c4c787b677c2d2a0ff0cbbca),[08803fa](https://gitlab.com/gocream/pysberbank/commit/08803fa510f6e8b9b5f2208132bed69137150f5f),[be3d329](https://gitlab.com/gocream/pysberbank/commit/be3d329735c1e38ea7d42e9c1b88e3a055b98c7d),[886a182](https://gitlab.com/gocream/pysberbank/commit/886a18256bdbf792ac8ff0777ee8bd9bf1d806c1),[1c00d39](https://gitlab.com/gocream/pysberbank/commit/1c00d39d57181de15feef040125855b84c44f4a2)]
- 🔧 (chore): Library refactor [[08d7545](https://gitlab.com/gocream/pysberbank/commit/08d754591e2def320d8b536f32dfef64c3110256),[2b47a22](https://gitlab.com/gocream/pysberbank/commit/2b47a224558cd2dfd24b1b99a5c61c45e35a56ab),[ebb257b](https://gitlab.com/gocream/pysberbank/commit/ebb257b595af8b85d25a0222fcdc5e03a86b589d),[f089c4f](https://gitlab.com/gocream/pysberbank/commit/f089c4f21e4718f57606d1eee6c4c5b83b2299b9)]


<a name="0.1.9"></a>
## 0.1.9 (2022-12-16)

### Miscellaneous

- 🐛 (client): Fixed encoding response [[b029105](https://gitlab.com/gocream/pysberbank/commit/b029105a910eeae375d8ba66aa608d175cf78bf2)]


<a name="0.1.8"></a>
## 0.1.8 (2022-12-13)

### Miscellaneous

- 🐛 (client): fixed order_status_extended method parameters [[a7e46ae](https://gitlab.com/gocream/pysberbank/commit/a7e46ae3a411fb251974339e6ccfbb2cbbbc5e86)]


<a name="0.1.7"></a>
## 0.1.7 (2022-12-02)

### Added

- ✨ (client): Default timeout for requests is 3 seconds for connect and 27 for read [[ea34b04](https://gitlab.com/gocream/pysberbank/commit/ea34b04e835e6fd5751dff863409f1dc5e2a8621)]
- ✨ (client): Pass requests parameters directly to the request [[7481b76](https://gitlab.com/gocream/pysberbank/commit/7481b7672102594993a1cec6fb71f19c3924ab15)]

### Miscellaneous

-  upate README [[7b5ea14](https://gitlab.com/gocream/pysberbank/commit/7b5ea14f11dedb0ceded0f501af7182dcc209001)]
-  typo [[cabd8f4](https://gitlab.com/gocream/pysberbank/commit/cabd8f49a2836e005fc1bf56d31066efa148c753)]
-  fix: Get rid of the dependency when installing the package [[7bbfda9](https://gitlab.com/gocream/pysberbank/commit/7bbfda97f4843b17f1b84ed3b2002503a37f732b)]
-  refactor: Added pyproject.toml for update installation configuration [[974b0a9](https://gitlab.com/gocream/pysberbank/commit/974b0a9ffa36db5207947959ff7b53c597b430f6)]
-  chore: Added undocumented error [[0415523](https://gitlab.com/gocream/pysberbank/commit/0415523c6b63fbad1e33bd6a5a19c91a672ff1a1)]
-  refactor: Package configuration refactor [[49a8e31](https://gitlab.com/gocream/pysberbank/commit/49a8e31b8d2acef9929c94b57dc94b601fe887d9)]


<a name="0.1.6"></a>
## 0.1.6 (2021-06-30)

### Added

- ✨ (constants): Added action codes [[2825c28](https://gitlab.com/gocream/pysberbank/commit/2825c28d1a52690b8533462321efc39fb6a94283)]


<a name="0.1.5"></a>
## 0.1.5 (2020-12-02)

### Miscellaneous

-  Change requests requirement to >=2.18.4,<3 [[187f5a4](https://gitlab.com/gocream/pysberbank/commit/187f5a41036d1d7338c595c7ab65e22a45b87b41)]
-  lint and more docs [[417d5db](https://gitlab.com/gocream/pysberbank/commit/417d5db9c23ef7b99d5a34dcc1256fb585515bd8)]
-  update gitignore [[37a1732](https://gitlab.com/gocream/pysberbank/commit/37a1732137ed693685135b0461b1114a34aeeb79)]


<a name="0.1.4"></a>
## 0.1.4 (2018-06-14)

### Miscellaneous

-  Fix sber API params [[7ff052d](https://gitlab.com/gocream/pysberbank/commit/7ff052d3840e3c668870daf0017e73def6e050cc)]


<a name="0.1.3"></a>
## 0.1.3 (2018-06-06)

### Miscellaneous

-  Added orderBundle and taxSystem to register() and version to 0.1.3 [[2e9267d](https://gitlab.com/gocream/pysberbank/commit/2e9267d8fc59fa7208d0061e174ebcf7cb792b13)]


<a name="0.1.2"></a>
## 0.1.2 (2018-06-05)

### Miscellaneous

-  Bump version [[d0d9d78](https://gitlab.com/gocream/pysberbank/commit/d0d9d7885ac3ff38d0c213f2dafd95ab7a835120)]
-  Formatting [[172df93](https://gitlab.com/gocream/pysberbank/commit/172df932502441ba8e8723cfd1bb3c7dc27a09b3)]
-  Formatting [[9613085](https://gitlab.com/gocream/pysberbank/commit/96130853e4c8c22d50888a4dce9c2b2a8219b42f)]
-  Reworked exceptions [[02cc2ae](https://gitlab.com/gocream/pysberbank/commit/02cc2ae6e08c318f4c45975e208c0065d3fe34c0)]


<a name="0.1.1"></a>
## 0.1.1 (2018-05-25)

### Miscellaneous

-  [update] version [[fcbe454](https://gitlab.com/gocream/pysberbank/commit/fcbe454534fb7306b733359167bbb02c69320d4c)]
-  [fix] install script [[7aaf702](https://gitlab.com/gocream/pysberbank/commit/7aaf702d625845f92235274cbc49504578f5a562)]


<a name="0.1.0"></a>
## 0.1.0 (2018-05-21)

### Miscellaneous

-  clear [[59a7889](https://gitlab.com/gocream/pysberbank/commit/59a78893dc7ad2b41d803b7eb1ffea91fa15b38a)]
-  order status extended tests [[c1acf5c](https://gitlab.com/gocream/pysberbank/commit/c1acf5cbdca8b7c65327940463e902665f94a7d0)]
-  update full status request and update errors [[b0a5d76](https://gitlab.com/gocream/pysberbank/commit/b0a5d760d3473192a624bdb3f869d37c96dfb4ce)]
-  order_status update + tests [[2ad7fa0](https://gitlab.com/gocream/pysberbank/commit/2ad7fa061e367bee7d5257d07ccb8f91093ec432)]
-  add refund tests [[a3e5000](https://gitlab.com/gocream/pysberbank/commit/a3e50009bb8cf46e9f1bf484be5fda4c01dccb7e)]
-  fix refund request [[2886668](https://gitlab.com/gocream/pysberbank/commit/2886668ad081ae3739d590d22e5c15c7bd3fe40e)]
-  update params in tests [[c886dd8](https://gitlab.com/gocream/pysberbank/commit/c886dd83c22c655220f7d98921c53a623c57cd5f)]
-  fix test and http + add reverse tests [[cde9b77](https://gitlab.com/gocream/pysberbank/commit/cde9b779b2e98d049cdeaca566de10dffb59669b)]
-  update structure of test folder [[9316345](https://gitlab.com/gocream/pysberbank/commit/9316345d5c4f9ecc130cc4e4681fe271f05f74f4)]
-  fix test. errorCode 0 is not error [[cacbb17](https://gitlab.com/gocream/pysberbank/commit/cacbb1701d5cbd09e6eb00781639db5804cc4f55)]

tests/refund.py
# -*- coding: utf-8 -*-

from unittest import TestCase
from unittest.mock import MagicMock
from unittest.mock import patch
import io
import json
import requests
import unittest

from pysberbank import Sberbank
import pysberbank



REQUEST = {
    'errorCode': "0",
    'errorMessage': "Успешно",
}
ERRORS = {
    "0": {
        "errorCode": "0",
        "errorMessage": "Обработка запроса прошла без системных ошибок.",
    },
    "5000": {
        "errorCode": "5",
        "errorMessage": "Доступ запрещён.",
    },
    "5001": {
        "errorCode": "5",
        "errorMessage": "Пользователь должен сменить свой пароль.",
    },
    "5002": {
        "errorCode": "5",
        "errorMessage": "[orderId] не задан.",
    },
    "5003": {
        "errorCode": "5",
        "errorMessage": "Неверная сумма.",
    },
    "6000": {
        "errorCode": "6",
        "errorMessage": "Неверный номер заказа.",
    },
    "7000": {
        "errorCode": "7",
        "errorMessage": "Платёж должен быть в корректном состоянии.",
    },
    "7001": {
        "errorCode": "7",
        "errorMessage": "Сумма возврата превышает сумму списания.",
    },
    "7002": {
        "errorCode": "7",
        "errorMessage": "Системная ошибка.",
    },
    '42': REQUEST
}


def _test_request(url, params=None, **kwargs):
    response = requests.Response()

    if url == "https://3dsec.sberbank.ru/payment/rest/refund.do":
        data = kwargs['data']

        userName = data.get('userName', None)
        password = data.get('password', None)
        orderId = data.get('orderId', None)

        if orderId in ERRORS:
            result = orderId
        else:
            result = '42'

        result = ERRORS[result]

    response.raw = io.BytesIO(str.encode(json.dumps(result)))
    return response


class RefundTestCase(TestCase):

    def setUp(self):
        self.pathed_requests = patch('requests.post', side_effect=_test_request)
        self.mock_requests = self.pathed_requests.start()

    def tearDown(self):
        self.pathed_requests.stop()

    def test_refund(self):
        api = Sberbank("test_username", "test_password")
        response = api.refund(
            orderId = "4302d369-a5e8-4432-a5e5-42acfab52c86",
            refundAmount = 20000,
            language = "ru",
        )
        self.assertIsNotNone(response)
        self.assertEqual(response, REQUEST)

    def test_refund_errors(self):
        api = Sberbank("test_username", "test_password")

        for k, v in ERRORS.items():
            with self.subTest(orderNumber=k):
                if k in ['42', '0']:
                    continue

                with self.assertRaises(pysberbank.errors.SberbankApiException):
                    response = api.refund(k, 20000)


if __name__ == '__main__':
    unittest.main()


tests/reverse.py
# -*- coding: utf-8 -*-

from unittest import TestCase
from unittest.mock import MagicMock
from unittest.mock import patch
import io
import json
import requests
import unittest

from pysberbank import Sberbank
import pysberbank



REVERSE_REQUEST = {
    'errorCode': "0",
    'errorMessage': "Успешно",
}
REVERSE_ERRORS = {
    '0': {
        'errorCode': "0",
        'errorMessage': "Обработка запроса прошла без системных ошибок.",
    },
    '5000': {
        'errorCode': "5",
        'errorMessage': "Доступ запрещён.",
    },
    '5001': {
        'errorCode': "5",
        'errorMessage': "Пользователь должен сменить свой пароль.",
    },
    '5002': {
        'errorCode': "5",
        'errorMessage': "[orderId] не задан.",
    },
    '6000': {
        'errorCode': "6",
        'errorMessage': "Неверный номер заказа.",
    },
    '7000': {
        'errorCode': "7",
        'errorMessage': "Недопустимая операция для текущего состояния заказа.",
    },
    '7001': {
        'errorCode': "7",
        'errorMessage': "Системная ошибка.",
    },
    '42': REVERSE_REQUEST
}


def _test_request(url, params=None, **kwargs):
    response = requests.Response()

    if url == "https://3dsec.sberbank.ru/payment/rest/reverse.do":
        # register order
        data = kwargs['data']

        userName = data.get('userName', None)
        password = data.get('password', None)
        orderId = data.get('orderId', None)

        if orderId in REVERSE_ERRORS:
            result = orderId
        else:
            result = '42'

        result = REVERSE_ERRORS[result]

    response.raw = io.BytesIO(str.encode(json.dumps(result)))
    return response


class ReverseTestCase(TestCase):

    def setUp(self):
        self.pathed_requests = patch('requests.post', side_effect=_test_request)
        self.mock_requests = self.pathed_requests.start()

    def tearDown(self):
        self.pathed_requests.stop()

    def test_register(self):
        api = Sberbank("test_username", "test_password")
        response = api.reverse(
            orderId = '9231a838-ac68-4a3e-bddb-d9781433d852',
            language = 'ru',
        )
        self.assertIsNotNone(response)
        self.assertEqual(response, REVERSE_REQUEST)

    def test_register_errors(self):
        api = Sberbank("test_username", "test_password")

        for k, v in REVERSE_ERRORS.items():
            with self.subTest(orderNumber=k):
                if k in ['42', '0']:
                    continue

                with self.assertRaises(pysberbank.errors.SberbankApiException):
                    response = api.reverse(k)


if __name__ == '__main__':
    unittest.main()


tests/__init__.py
# -*- coding: utf-8 -*-

import unittest

from .register import RegisterTestCase
from .reverse import ReverseTestCase
from .refund import RefundTestCase
from .order_status import OrderStatusTestCase
from .order_status_extended import OrderStatusExtendedTestCase


def suite():
    suite = unittest.TestSuite()
    suite.addTest(RegisterTestCase('test_register'))
    suite.addTest(RegisterTestCase('test_register_errors'))
    suite.addTest(ReverseTestCase('test_reverse'))
    suite.addTest(ReverseTestCase('test_reverse_errors'))
    suite.addTest(RefundTestCase('test_refund'))
    suite.addTest(RefundTestCase('test_refund_errors'))
    suite.addTest(OrderStatusTestCase('test_order_status'))
    suite.addTest(OrderStatusTestCase('test_order_status_errors'))
    suite.addTest(OrderStatusExtendedTestCase('order_status_extended'))
    suite.addTest(OrderStatusExtendedTestCase('order_status_extended_errors'))
    return suite


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())


pysberbank/http.py
# -*- coding: utf-8 -*-

from requests import RequestException
import json
import logging
import requests

from .errors import SberbankApiException
from .errors import SberbankException
from .errors import SberbankRequestException


logger = logging.getLogger("pysberbank")


DEFAULT_URL = "https://3dsec.sberbank.ru/payment/{api_type}/{method}"
DEFAULT_API_TYPE = "rest"
DEFAULT_LANGUAGE = "ru"
DEFAULT_CURREBCY = 643


class SberbankPaymentApi:
    """
        sberbank payment api
    """

    def __init__(self, username=None, password=None, token=None,
                 url=DEFAULT_URL, api_type=DEFAULT_API_TYPE,
                 language=DEFAULT_LANGUAGE, currency=DEFAULT_CURREBCY,
                 ensure_ascii=True):
        """
            username - username at sberbank api
            password - password at sberbank api
            url - api url address
            api_type - api type (soap | rest) only rest ready
            language - default language (ISO 639-1)
            currency - default currency (ISO 4217)

            @TODO:
             * remove `url` parameter in favor of a boolean `production` parameter
             * move `ensure_ascii` parameter to `_make_data` method parameter and pass it from class methods
        """

        if (not (bool(username) and bool(password))) == (token is None):
            raise SberbankException(
                "Auth params require. "
                "Setup username/password pair or token, not both"
            )

        self.username = username
        self.password = password
        self.token = token

        self.url = url
        self.api_type = api_type
        self.language = language
        self.currency = currency

        self.ensure_ascii = ensure_ascii

    def _make_data(self, params, remove_null=True):
        """
        add username/password pair or token to request params

        remove_null: if True - remove all None elements from params
        """

        if remove_null:
            data = dict(filter(lambda x: x[1], params.items()))
        else:
            data = params.copy()

        if self.token:
            data['token'] = self.token
        else:
            data.update({
                'userName': self.username,
                'password': self.password,
            })

        # we need to encode lists and dicts in json
        for k, v in data.items():
            if not isinstance(v, (list, dict)):
                continue
            data[k] = json.dumps(data[k], ensure_ascii=self.ensure_ascii)

        return data

    def _make_request(
        self,
        url,
        api_type,
        method_name,
        params=None,
        remove_null=True,
        **kwargs,
    ):
        # resource
        if api_type == 'rest':
            method = f'{method_name}.do'
        else:
            method = method_name
        url = url.format(api_type=api_type, method=method)

        # pyload
        if params is None:
            params = {}
        data = self._make_data(params, remove_null=remove_null)

        # request params
        kwargs.setdefault('timeout', (3, 27))

        logger.debug(f"Make request to {url} with data {data} and params {kwargs}.")

        try:
            response = requests.post(url, data=data, **kwargs)
            logger.debug(f"Response to {url} request is {response.content}.")

            if api_type == 'rest':
                result = response.json()

                if result.get('errorCode', '0') != '0':
                    raise SberbankApiException(result['errorCode'],
                                               result['errorMessage'])
            else:
                result = response.content

        except RequestException as e:
            # wrap Requests exception in our for filtration purposes
            raise SberbankRequestException(
                request=e.request,
                response=e.response
            ) from e

        return result

    def register(
        self,
        orderNumber,
        amount,
        returnUrl,
        currency=None,
        failUrl=None,
        description=None,
        language=None,
        pageView=None,
        clientId=None,
        merchantLogin=None,
        jsonParams=None,
        order_bundle=None,
        tax_system=None,
        sessionTimeoutSecs=None,
        expirationDate=None,
        bindingId=None,
        features=None,
        url=None,
        api_type=None,
        **kwargs,
    ):
        """
        Register cart

        `Request doc`_
        `Request Cart doc`_

        Attributes:
            orderBundle: корзина для чека
            taxSystem: система налогообложения

        .. _Request doc:
           https://securepayments.sberbank.ru/wiki/doku.php/integration:api:rest:requests:register

        .. _Request Cart doc:
           https://securepayments.sberbank.ru/wiki/doku.php/integration:api:rest:requests:register_cart
        """

        params = {
            'orderNumber': orderNumber,
            'amount': amount,
            'currency': currency or self.currency,
            'returnUrl': returnUrl,
            'failUrl': failUrl,
            'description': description,
            'language': language or self.language,
            'pageView': pageView,
            'clientId': clientId,
            'merchantLogin': merchantLogin,
            'jsonParams': jsonParams,
            'sessionTimeoutSecs': sessionTimeoutSecs,
            'expirationDate': expirationDate,
            'bindingId': bindingId,
            'features': features,
            'orderBundle': order_bundle,
            'taxSystem': tax_system,
        }
        result = self._make_request(url or self.url, api_type or self.api_type,
                                    'register', params, **kwargs)
        return result

    def reverse(self, orderId, language=None, url=None, api_type=None, **kwargs):
        """
        Reverse payment

        Функция отмены доступна в течение ограниченного времени после оплаты,
        точные сроки необходимо уточнять в «Сбербанке».

        `Request doc`_

        Attributes:
            orderId: Номер заказа в платёжном шлюзе. Уникален в пределах шлюза.

        .. _Request doc:
           https://securepayments.sberbank.ru/wiki/doku.php/integration:api:rest:requests:reverse
        """

        params = {
            'orderId': orderId,
            'language': language or self.language,
        }
        result = self._make_request(url or self.url, api_type or self.api_type,
                                    'reverse', params, **kwargs)
        return result

    def refund(self, orderId, refundAmount, language=None, url=None, api_type=None, **kwargs):
        """
        Refund

        `Request doc`_

        Attributes
            orderId: Номер заказа в платёжном шлюзе. Уникален в пределах шлюза.
            refundAmount: Сумма возврата в валюте заказа. Может быть меньше или
                равна остатку в заказе.
            language: Не указан в документации, но вероятно по аналогии с
                остальными методами он всё же там должен быть... Вероятно...
                Нужно поправить если ругаться будет

        .. _Request doc:
           https://securepayments.sberbank.ru/wiki/doku.php/integration:api:rest:requests:refund
        """

        params = {
            'orderId': orderId,
            'amount': refundAmount,
            'language': language or self.language,
        }
        result = self._make_request(url or self.url, api_type or self.api_type,
                                    'refund', params, **kwargs)
        return result

    def order_status(self, orderId, language=None, url=None, api_type=None, **kwargs):
        """
        Order status

        `Request doc`_

        Attributes
            orderId: Номер заказа в платёжном шлюзе. Уникален в пределах шлюза.
            language: Не указан в документации, но вероятно по аналогии с
                остальными методами он всё же там должен быть... Вероятно...
                Нужно поправить если ругаться будет

        .. _Request doc:
           https://securepayments.sberbank.ru/wiki/doku.php/integration:api:rest:requests:getorderstatus
        """

        params = {
            'orderId': orderId,
            'language': language or self.language,
        }
        result = self._make_request(url or self.url, api_type or self.api_type,
                                    'getOrderStatus', params, **kwargs)
        return result

    def order_status_extended(
        self,
        orderId=None,
        orderNumber=None,
        anguage=None,
        url=None,
        api_type=None,
        language=None,
        **kwargs,
    ):
        """
        Order status extended

        В запросе должен присутствовать либо _orderId_, либо _orderNumber_.
        Если в запросе присутствуют оба параметра, то приоритетным считается
        _orderId_.

        `Request doc`_

        Attributes
            orderId: Номер заказа в платёжном шлюзе. Уникален в пределах шлюза.
            orderNumber: Номер (идентификатор) заказа в системе магазина,
            уникален для каждого магазина в пределах системы - до 30 символов.
            language: Не указан в документации, но вероятно по аналогии с
                остальными методами он всё же там должен быть... Вероятно...
                Нужно поправить если ругаться будет

        .. _Request doc:
           https://securepayments.sberbank.ru/wiki/doku.php/integration:api:rest:requests:getorderstatusextended
        """

        if (orderId is None) and (orderNumber is None):
            raise SberbankException(
                "none of the orderId and "
                "orderNumber parameters is specified"
            )

        params = {
            'orderId': orderId,
            'orderNumber': orderNumber,
            'language': language or self.language,
        }
        result = self._make_request(url or self.url, api_type or self.api_type,
                                    'getOrderStatusExtended', params, **kwargs)
        return result
